staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/matchconditions/matcher.go

```diff
@@ -0,0 +1,144 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package matchconditions
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/google/cel-go/cel"
+	celtypes "github.com/google/cel-go/common/types"
+
+	v1 "k8s.io/api/admissionregistration/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apiserver/pkg/admission"
+	admissionmetrics "k8s.io/apiserver/pkg/admission/metrics"
+	celplugin "k8s.io/apiserver/pkg/admission/plugin/cel"
+	celconfig "k8s.io/apiserver/pkg/apis/cel"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	"k8s.io/klog/v2"
+)
+
+var _ celplugin.ExpressionAccessor = &MatchCondition{}
+
+// MatchCondition contains the inputs needed to compile, evaluate and match a cel expression
+type MatchCondition v1.MatchCondition
+
+func (v *MatchCondition) GetExpression() string {
+	return v.Expression
+}
+
+func (v *MatchCondition) ReturnTypes() []*cel.Type {
+	return []*cel.Type{cel.BoolType}
+}
+
+var _ Matcher = &matcher{}
+
+// matcher evaluates compiled cel expressions and determines if they match the given request or not
+type matcher struct {
+	filter      celplugin.Filter
+	authorizer  authorizer.Authorizer
+	failPolicy  *v1.FailurePolicyType
+	matcherType string
+	objectName  string
+}
+
+func NewMatcher(filter celplugin.Filter, authorizer authorizer.Authorizer, failPolicy *v1.FailurePolicyType, matcherType, objectName string) Matcher {
+	return &matcher{
+		filter:      filter,
+		authorizer:  authorizer,
+		failPolicy:  failPolicy,
+		matcherType: matcherType,
+		objectName:  objectName,
+	}
+}
+
+func (m *matcher) Match(ctx context.Context, versionedAttr *admission.VersionedAttributes, versionedParams runtime.Object) MatchResult {
+	var f v1.FailurePolicyType
+	if m.failPolicy == nil {
+		f = v1.Fail
+	} else {
+		f = *m.failPolicy
+	}
+
+	evalResults, err := m.filter.ForInput(ctx, versionedAttr, celplugin.CreateAdmissionRequest(versionedAttr.Attributes), celplugin.OptionalVariableBindings{
+		VersionedParams: versionedParams,
+		Authorizer:      m.authorizer,
+	}, celconfig.RuntimeCELCostBudgetMatchConditions)
+
+	//TODO: add event to webhook object for error
+	if err != nil {
+		// filter returning error is unexpected and not an evaluation error so not incrementing metric here
+		if f == v1.Fail {
+			return MatchResult{
+				Error: err,
+			}
+		} else {
+			return MatchResult{
+				Matches: true,
+			}
```

Comment:
Yeah I created a tracking issue for this as graduation criteria https://github.com/kubernetes/kubernetes/issues/116609

Reply:
given @maxsmythe's comment, I think a condition eval error should skip the webhook call for `failurePolicy: ignore` in alpha. @maxsmythe / @tallclair - thoughts?

Reply:
IMO "good enough" behavior meets alpha requirements (which assume future changes and the existence of flaws), and def. don't want to block if there is debate.

I'm fine with whatever behavior for alpha so long as we can fully weigh the pros/cons before beta

Reply:
I don't see how that changes the consideration for the failure policy? I still think `failurePolicy: ignore` defaulting to calling the webhook (if no expressions returned false) is the right behavior.

Reply:
because it means injecting an expression which always evaluates to false is not sufficient to be sure the webhook is not called (because that expression might never get evaluated because the budget was exhausted)

Reply:
If I wanted to force a webhook to be called regardless of other match conditions:

* Set failurePolicy: ignore
* Create an arbitrarily expensive match condition
* Have it evaluate first (or wait for random evaluation order to work out)
* All later webhooks are ignored

In the other thread you mentioned a mitigation of having providers rely on forcing order (though this may be hard to do with merge strategies).

I don't think we want to try to game this out in the next 2.5 hrs, so am happy to go with a reasonable, if flawed, solution for first alpha and require a more considered approach before beta.

Reply:
Yeah, I agree with just picking something for alpha and blocking the beta on a more complete strategy for this.

Reply:
Maybe add a TODO here, and link to the issue you filed.